README.md
# hubot-phrases

[![Build Status](https://travis-ci.org/halkeye/hubot-phrases.png?branch=master)](https://travis-ci.org/halkeye/hubot-phrases)
[![Codacy Badge](https://api.codacy.com/project/badge/Grade/667ad904c0dd4cea94f50e513720e71a)](https://www.codacy.com/app/halkeye/hubot-phrases?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=halkeye/hubot-phrases&amp;utm_campaign=Badge_Grade)
[![Codacy Badge](https://api.codacy.com/project/badge/Coverage/667ad904c0dd4cea94f50e513720e71a)](https://www.codacy.com/app/halkeye/hubot-phrases?utm_source=github.com&utm_medium=referral&utm_content=halkeye/hubot-phrases&utm_campaign=Badge_Coverage)

phrases (responses) support for hubot

## Getting Started
1. Install the module: `npm install --save hubot-phrases`
2. Add it `hubot-phrases` to your external-scripts.json file in your hubot directory

## Usage

* hubot literal <phrase> - Outputs information about a phrase. if greater than 10 tidbits, then output a url instead
* hubot do something|something random - output a random phrase
* hubot forget <phrase>#<num> - Deletes the <num>th response of <phrase>
* hubot forget that - Deletes the last used response
* hubot what was that - Outputs what the last phrase was
* hubot X <reply>|<action>|is Y - Sets a response for X to be Y
* [phrase] - Outputs a random tidbit/reply that matches <phrase>

## Configuration

None

## Release History

2016-12-30 - 1.0.2

* Minor bug fix to handle when no existing data

2016-12-27 - 1.0.1

* Full cleanup/rename of factoid to phrases
* Update documentation

2016-12-27 - 1.0.0

* Remove coffeescript
* Test coverage
* minor bug fixes

## License
Copyright (c) 2013,2016 Gavin Mogan
Licensed under the MIT license.


CHANGELOG.md
<a name="1.0.2"></a>
## [1.0.2](https://github.com/halkeye/hubot-phrases/compare/v1.0.1...v1.0.2) (2016-12-30)



<a name="1.0.1"></a>
## [1.0.1](https://github.com/halkeye/hubot-phrases/compare/v1.0.0...v1.0.1) (2016-12-27)



<a name="1.0.0"></a>
# [1.0.0](https://github.com/halkeye/hubot-phrases/compare/f21a781...v1.0.0) (2016-12-27)


### Chores

* drop support for Node.js 0.10 ([f21a781](https://github.com/halkeye/hubot-phrases/commit/f21a781))


### BREAKING CHANGES

* This module no longer supports Node.js 0.10





Gruntfile.js
module.exports = function(grunt) {
  'use strict';

  // Project configuration.
  grunt.initConfig({
    pkg: grunt.file.readJSON('package.json'),
    eslint: {
      target: ['src/**/*.js'],
    },
    simplemocha: {
      all: {
        src: [
          'node_modules/should/should.js',
          'src/test/**/*.js'
        ],
        options: {
          globals: ['should'],
          timeout: 3000,
          ignoreLeaks: false,
          //grep: '**/*.js'
          ui: 'bdd',
          bail: true,
          //reporter: 'tap'
          reporter: 'spec'
        }
      }
    },
    watch: {
      gruntfile: {
        files: '<%= eslint.gruntfile.src %>',
        tasks: ['eslint:gruntfile']
      },
      coffeeLib: {
        files: '<%= eslint.scripts.src %>',
        tasks: ['eslint:scripts', 'simplemocha']
      },
      coffeeTest: {
        files: ['<%= eslint.test.src %>','node_modules/hubot-variables/src/scripts/*.coffee'],
        tasks: [ 'eslint:test', 'simplemocha']
      }
    },
    clean: ['out/']});

  // plugins.
  grunt.loadNpmTasks('grunt-simple-mocha');
  grunt.loadNpmTasks('grunt-eslint');
  grunt.loadNpmTasks('grunt-contrib-clean');
  grunt.loadNpmTasks('grunt-contrib-watch');

  // tasks.
  grunt.registerTask('compile', [ 'eslint' ]);

  grunt.registerTask('test', [ 'simplemocha' ]);
  return grunt.registerTask('default', ['compile', 'test']);
};



src/scripts/hubot-factoid.js
// Description:
//   factoid (responses) support for hubot
//
// Dependencies:
//   None
//
// Configuration:
//   HUBOT_URL - Base url instead of hostname:port, ex: "http://hubot.kodekoan.com"
//
// URLS:
//   /hubot/factoid/:factoid
//
// Commands:
//   hubot literal <factoid> - Outputs information about a factoid. Right now only the url
//   hubot do something - see something random
//   hubot something random - see somethin random
//   do something - see something random
//   something random - Outputs a random factoid
//   hubot forget <factoid>#<num> - Deletes the <num>th response of <factoid>
//   hubot forget that - Deletes the last used response
//   hubot what was that - Outputs what the last factoid was
//   <factoid> - Outputs a random reply that matches <factoid>
//
// Notes:
//   Copyright (c) 2013 Gavin Mogan
//   Licensed under the MIT license.
//
// Author:
//   halkeye

const util = require('util');
const os = require('os');

module.exports = function Plugin (robot) {
  class Factoid {
    constructor (name, data) {
      if (!data) {
        data = {};
      }
      this.name = name;
      this.tidbits = [];
      this.alias = false;
      this.readonly = true;
      if ((data.readonly != null) && data.readonly === false) {
        this.readonly = false;
      }
      if (data.alias) {
        this.alias = data.alias;
      } else if (data.tidbits) {
        for (let tid of data.tidbits) {
          this.tidbits.push(tid);
        }
      }
    }
    canAlias (user) {
      // FIXME - op only commands according to bucket
      return this.canEdit(user);
    }
    canEdit (user) {
      if (user.roles) {
        if (user.roles.includes('edit_factoids')) {
          return true;
        }
        if (user.roles.includes(`edit_factoid_${this.name}`)) {
          return true;
        }
      }
      if (this.readonly) {
        return false;
      }
      return true;
    }
    tidbit () {
      return this.tidbits[Math.floor(Math.random() * this.tidbits.length)];
    }
    addTidbit (tidbit, verb) {
      // FIXME - add validation
      return this.tidbits.push({
        tidbit,
        verb
      });
    }

    toObj () {
      if (this.alias) {
        return {
          alias: this.alias,
          readonly: this.readonly
        };
      }
      return {
        readonly: this.readonly,
        tidbits: this.tidbits
      };
    }
    save () {
      const factoid = robot.brain.data.factoids[this.name] = this.toObj();
      if (!factoid.alias && !factoid.tidbits.length) {
        delete robot.brain.data.factoids[this.name];
      }
    }
  }

  class FactoidHandler {
    constructor () {
      this.alias = this.alias.bind(this);
      this.setAddressed = this.setAddressed.bind(this);
      this.set = this.set.bind(this);
      this.stats = {};
      this.facts = {};
      this.last_factoid = {};
      robot.brain.on('loaded', data => {
        this.facts = {};
        if (robot.brain.data.factoids) {
          robot.logger.info('Loading saved factoids');
          let keys = Object.keys(data.factoids);
          for (let key of keys) {
            this.facts[key] = new Factoid(key, data.factoids[key]);
          }
        }
        return robot.brain.emit('finished_loading_factoids');
      }
      );
    }
    hasFacts () {
      return Object.keys(this.facts).length;
    }
    get (str, history) {
      if (!this.hasFacts()) { return; }
      let factoid = this.facts[str];
      if (!factoid) { return; }
      robot.logger.debug(util.inspect(factoid));
      if (history) { history.push(factoid); }
      if (factoid.alias) { return this.get(factoid.alias, history); }
      return factoid;
    }
    random (history) {
      if (!this.hasFacts()) { return; }
      let keys = Object.keys(this.facts);
      let factoid = keys[Math.floor(Math.random() * keys.length)];
      return this.get(factoid, history);
    }
    output (msg, factoid, history) {
      let tidbit = factoid.tidbit();
      let outputHistory = { factoid, tidbit };
      // FIXME this should be per room
      if (history) { history.push(outputHistory); }
      let output = tidbit.tidbit;
      if (robot.variables != null) {
        output = robot.variables.process(output, msg.message.user, outputHistory);
      }
      if (tidbit.verb === '<reply>') {
        return msg.send(output);
      } else if (tidbit.verb === '<action>') {
        if (msg.emote != null) {
          return msg.emote(output);
        } else {
          return msg.send(`/me ${output}`);
        }
      } else if (tidbit.verb === 'is' && factoid.name.toLowerCase() === robot.name.toLowerCase()) {
        return msg.send(`I am ${output}`);
      } else {
        return msg.send([factoid.name, tidbit.verb, output].join(' '));
      }
    }
    alias (msg) {
      let srcName = msg.match[1].trim();
      let targetName = msg.match[2].trim();
      let src = robot.factoid.get(srcName);
      if (src) {
        msg.reply(`Sorry, there is already a factoid for '${srcName}'.`);
        return;
      }
      let target = robot.factoid.get(targetName);
      if (!target) {
        msg.reply(`Sorry, there is no factoid for the target '${targetName}'.`);
        return;
      }
      if (!target.canAlias(msg.message.user)) {
        robot.logger.debug(`${factoid} that factoid is protected`);
        msg.reply('Sorry, that factoid is protected');
        return;
      }
      msg.finish();
      robot.logger.info(`${msg.message.user.name} aliased '${srcName}' to '${targetName}'`);
      var factoid = this.facts[srcName] = new Factoid(srcName);
      factoid.alias = targetName;
      factoid.save();
      return msg.reply('Okay.');
    }
    setAddressed (msg) {
      msg.message.addressed = true;
      return this.set(msg);
    }
    set (msg) {
      msg.finish(); // we are adding a message, so ignore any other handler type
      let fact = msg.match[1].trim();
      let verb = msg.match[2].trim();
      let tidbit = msg.match[3].trim();
      let forced = !!msg.match[4];
      if (!msg.message.addressed && /^[^a-zA-Z]*<.?\S+>/.test(fact)) {
        robot.logger.debug('Not learning from what seems to be an IRC quote: $fact');
        return;
      }
      if (!msg.message.addressed && !forced && /=~/.test(fact)) {
        robot.logger.debug('Not learning what looks like a botched =~ query');
        msg.reply('Fix your =~ command.');
        return;
      }

      if (fact === 'you' && verb === 'are') {
        fact = robot.name;
        verb = 'is';
      } else if (fact === 'I' && verb === 'am') {
        fact = msg.message.user.name;
        verb = 'is';
      }

      this.stats.learn++;
      let matches = tidbit.match(/^<(action|reply)>\s*(.*)/);
      var also;
      if (matches) {
        verb = `<${matches[1]}>`;
        tidbit = matches[2];
      } else if (verb === 'is also') {
        also = 1;
        verb = 'is';
      } else if (forced) {
        if (verb !== '<action>' && verb !== '<reply>') {
          verb = verb.replace(/^<|>$/, '');
        }
        if (/\sis\salso$/.test(fact)) {
          also = 1;
        } else {
          fact.replace(/\sis$/, '');
        }
      }

      if (verb.toLowerCase === '<alias>') {
        msg.reply("please use the 'alias' command.");
        return;
      }

      fact = fact.trim();
      robot.logger.debug(`Learning ${fact} ${verb} ${tidbit}`);

      if (fact.toLowerCase() === msg.message.user.name.toLowerCase() || fact.toLowerCase() === msg.message.user.name.toLowerCase() + ' quotes') {
        robot.logger.debug(`Not allowing ${msg.message.user.name} to edit his own factoid`);
        msg.reply("Please don't edit your own factoids.");
        return;
      }

      let factoid = this.get(fact);
      if (!factoid) {
        factoid = this.facts[fact] = new Factoid(fact);
      } else if (!factoid.canEdit(msg.message.user)) {
        robot.logger.debug(`${factoid} that factoid is protected`);
        msg.reply('Sorry, that factoid is protected');
        return;
      }

      for (let t of factoid.tidbits) {
        if (tidbit.toLowerCase() === t.tidbit.toLowerCase()) {
          msg.reply('I already had it that way');
          return;
        }
      }

      factoid.addTidbit(tidbit, verb);
      factoid.tidbits.push;
      factoid.save();
      robot.logger.debug(`${msg.message.user.name} taught in ${msg.message.user.room} ${factoid.tidbits.length} '${fact}', '${verb}' '${tidbit}'`);
      return msg.reply('Okay.');
    }
    handlerLiteral (msg) {
      // page - http://wiki.xkcd.com/irc/bucket#Listing_factoids
      let page = msg.match[1];
      let factoidName = msg.match[2].trim();
      let factoid = robot.factoid.get(factoidName);
      if (!factoid) {
        msg.reply('No such factoid');
        return;
      }
      msg.finish();
      if (factoid.tidbits.length > 10) {
        let baseurl = (process.env.HUBOT_URL || (`http://${os.hostname()}:${robot.server.address().port}`)).replace(/\/+$/, '');
        msg.reply(baseurl + '/' + robot.name + '/factoid/' + encodeURIComponent(factoidName));
        return;
      }
      let response = [];
      factoid.tidbits.forEach(tidbit => response.push(`${tidbit.verb} ${tidbit.tidbit}`));
      return msg.reply(`${factoid.name} (${factoid.tidbits.length}): ${response.join('|')}`);
    }
  }

  robot.factoid = new FactoidHandler();
  robot.factoid.last_factoid = null;

  robot.respond(/(?:do something|something random)$/, msg => {
    let history = [];
    let factoid = robot.factoid.random(history);
    robot.factoid.output(msg, factoid, history);
    if (history.length > 0) { robot.factoid.last_factoid = history; }
  });

  robot.hear(/^(?:do something|something random)$/, msg => {
    let history = [];
    let factoid = robot.factoid.random(history);
    robot.factoid.output(msg, factoid, history);
    if (history.length > 0) { robot.factoid.last_factoid = history; }
  });

  robot.respond(/(un)?protect\s*(.*)$/, msg => {
    let protect = !msg.match[1];
    let factoidName = msg.match[2].trim();
    let factoid = robot.factoid.get(factoidName);
    if (!factoid) {
      msg.reply('No such factoid');
      return;
    }
    msg.finish();
    if (factoid.readonly === protect) {
      msg.reply('I already had it that way');
      return;
    }
    factoid.readonly = protect;
    factoid.save();
    return msg.reply('Okay.');
  }
  );

  robot.respond(/alias (.*?) => (.*?)$/, robot.factoid.alias);

  robot.router.get(`/${robot.name}/factoid/:factoid`, function (req, res) {
    let factoidName = req.params.factoid;
    let factoid = robot.factoid.get(factoidName);
    if (!factoid) { return res.status(404).send('Not Found'); }
    res.setHeader('content-type', 'text/plain');
    let content = [];
    content.push(`Factoid: [${factoidName}]`);
    content.push(`Protected: ${factoid.readonly ? 'true' : 'false'}`);
    content.push('');
    content.push('Tidbits:');
    for (let tidbit of factoid.tidbits) {
      content.push(tidbit.verb + '|' + tidbit.tidbit);
    }
    res.send(content.join('\n'));
    return res.end;
  });
    // res.send require('util').inspect(req.params)

  robot.respond(/literal(?:\[([*\d]+)])?\s+(.*)$/, robot.factoid.handlerLiteral);
  robot.hear(/^literal(?:\[([*\d]+)])?\s+(.*)$/, robot.factoid.handlerLiteral);

  robot.respond(/forget #(\d+)$/, msg => {
    return msg.reply('Sorry, syntax is now "forget <factoid>#<index from 0>" or "forget that"');
  });

  robot.respond(/forget that$/, msg => {
    return msg.reply('FIXME - not implemented yet');
  });

  robot.respond(/forget (.+)#(\d+)$/, msg => {
    let factoidName = msg.match[1].trim();
    let tid = parseInt(msg.match[2], 10) - 1;
    let factoid = robot.factoid.get(factoidName);
    if (!factoid) {
      msg.reply('No such factoid');
      return;
    }
    if (!factoid.canEdit(msg.message.user)) {
      msg.reply(`Sorry, you don't have permissions to edit '${factoid.name}'.`);
      return;
    }
    if (tid < 0) {
      msg.reply('Sorry, you must provide a number greater than 0 (as this is 1 based)');
      return;
    }
    if (!factoid.tidbits[tid]) {
      msg.reply(`Can't find tidbit #${tid + 1}`);
      return;
    }
    let tidbit = factoid.tidbits.splice(tid, 1);
    msg.reply(`Deleted tidbit: ${tidbit[0].verb}|${tidbit[0].tidbit}`);
    return factoid.save();
  });

  robot.respond(/what was that\??$/, function (msg) {
    // FIXME this should be per room
    if (!robot.factoid.last_factoid) { return; }
    msg.finish();
    // halkeye: That was 'rofl' (#315): <reply> I am also amused
    // halkeye: That was 'that's what she said' => 'thats what she said' (#65): <reply> No, that's what HE said.
    // halkeye: That was 'give me a weapon' (#863): <action> gives $weapon to $who;  vars used: { 'weapon' => [ 'a Biggoron Sword' ]};.
    let lf = robot.factoid.last_factoid.slice(-1)[0];
    let { name } = lf.factoid;
    let { tidbit } = lf;
    let idx = lf.factoid.tidbits.map(tid => tid.tidbit).indexOf(tidbit.tidbit);

    let response = [];
    response.push('That was');
    if (robot.factoid.last_factoid.length > 2) {
      robot.factoid.last_factoid.slice(0, -2).forEach(fact => response.push(`'${fact.name}' =>`));
    }
    response.push(`'${name}'`);
    response.push(`(#${idx}):`);
    response.push(tidbit.verb);
    response.push(tidbit.tidbit);
    if (lf.vars) {
      response.push(';');
      response.push('vars used:');
      response.push(util.inspect(lf.vars, { depth: null }));
    }
    return msg.reply(response.join(' '));
  });

  robot.respond(/(.*?)\s+(<\w+(?:'t)?>)\s*(.*)()/i, robot.factoid.setAddressed);
  robot.respond(/(.*?)(<'s>)\s+(.*)()/i, robot.factoid.setAddressed);
  robot.respond(/(.*?)\s+(is(?: also)?|are)\s+(.*)/i, robot.factoid.setAddressed);
  robot.hear(/^(.*?)\s+(<\w+(?:'t)?>)\s*(.*)()/i, robot.factoid.set);
  robot.hear(/^(.*?)(<'s>)\s+(.*)()/i, robot.factoid.set);
  robot.hear(/^(.*?)\s+(is(?: also)?|are)\s+(.*)/i, robot.factoid.set);

  // # FIXME, make these loaded once brain is loaded so it doesn't need to do wildcard match
  return robot.hear(/^(.*)\??$/, msg => {
    let factoidName = msg.match[1].trim();
    // FIXME this should be per room
    let history = [];
    let factoid = robot.factoid.get(factoidName, history);
    if (!factoid) { return; }
    robot.factoid.output(msg, factoid, history);
    if (history.length > 0) { robot.factoid.last_factoid = history; }
  });
};



src/scripts/hubot-phrases.js
// Description:
//   phrase (responses) support for hubot
//
// Dependencies:
//   None
//
// Configuration:
//   HUBOT_URL - Base url instead of hostname:port, ex: "http://hubot.kodekoan.com"
//
// URLS:
//   /hubot/phrase/:phrase
//
// Commands:
//    hubot literal <phrase> - Outputs information about a phrase. if greater than 10 tidbits, then output a url instead
//    hubot do something|something random - output a random phrase
//    hubot forget <phrase>#<num> - Deletes the <num>th response of <phrase>
//    hubot forget that - Deletes the last used response
//    hubot what was that - Outputs what the last phrase was
//    hubot X <reply>|<action>|is Y - Sets a response for X to be Y
//    <phrase> - Outputs a random tidbit/reply that matches <phrase>
//
// Notes:
//   Copyright (c) 2013,2016 Gavin Mogan
//   Licensed under the MIT license.
//
// Author:
//   halkeye

const util = require('util');
const os = require('os');

module.exports = function Plugin (robot) {
  class Factoid {
    constructor (name, data) {
      if (!data) {
        data = {};
      }
      this.name = name;
      this.tidbits = [];
      this.alias = false;
      this.readonly = false;
      if (data.readonly === true) {
        this.readonly = data.readonly;
      }
      if (data.alias) {
        this.alias = data.alias;
      } else if (data.tidbits) {
        for (let tid of data.tidbits) {
          this.tidbits.push(tid);
        }
      }
    }
    canAlias (user) {
      // FIXME - op only commands according to bucket
      return this.canEdit(user);
    }
    canEdit (user) {
      if (user.roles) {
        if (user.roles.includes('edit_phrases')) {
          return true;
        }
        if (user.roles.includes(`edit_phrase_${this.name}`)) {
          return true;
        }
      }
      if (this.readonly) {
        return false;
      }
      return true;
    }
    tidbit () {
      return this.tidbits[Math.floor(Math.random() * this.tidbits.length)];
    }

    toObj () {
      if (this.alias) {
        return {
          alias: this.alias,
          readonly: this.readonly
        };
      }
      return {
        readonly: this.readonly,
        tidbits: this.tidbits
      };
    }
    save () {
      const phrases = robot.brain.get('phrases') || {};
      const phrase = phrases[this.name] = this.toObj();
      if (!phrase.alias && !phrase.tidbits.length) {
        delete phrases[this.name];
      }
      robot.brain.set('phrases', phrases);
    }
  }

  class FactoidHandler {
    constructor () {
      this.alias = this.alias.bind(this);
      this.setAddressed = this.setAddressed.bind(this);
      this.set = this.set.bind(this);
      this.handlerGetAddressed = this.handlerGetAddressed.bind(this);
      this.handlerGet = this.handlerGet.bind(this);

      this.stats = {};
      this.last_phrase = {};
      robot.brain.on('loaded', (data = {_private: {}}) => {
        robot.brain.emit('finished_loading_phrases');
      });
    }
    hasFacts () {
      return Object.keys(robot.brain.get('phrases') || {}).length;
    }
    get (str, history) {
      if (!this.hasFacts()) { return; }
      let phrases = robot.brain.get('phrases');
      if (!phrases || !phrases[str]) { return; }
      let phrase = new Factoid(str, phrases[str]);
      if (!phrase) { return; }
      robot.logger.debug(util.inspect(phrase));
      if (history) { history.push(phrase); }
      if (phrase.alias) { return this.get(phrase.alias, history); }
      return phrase;
    }
    random (history) {
      if (!this.hasFacts()) { return; }
      let keys = Object.keys(robot.brain.get('phrases'));
      let phrase = keys[Math.floor(Math.random() * keys.length)];
      return this.get(phrase, history);
    }
    output (msg, phrase, history) {
      let tidbit = phrase.tidbit();
      let outputHistory = { phrase, tidbit };
      // FIXME this should be per room
      if (history) { history.push(outputHistory); }
      let output = tidbit.tidbit;
      if (robot.variables != null) {
        output = robot.variables.process(output, msg.message.user, outputHistory);
      }
      if (tidbit.verb === '<reply>') {
        return msg.send(output);
      } else if (tidbit.verb === '<action>') {
        if (msg.emote != null) {
          return msg.emote(output);
        } else {
          return msg.send(`/me ${output}`);
        }
      } else if (tidbit.verb === 'is' && phrase.name.toLowerCase() === robot.name.toLowerCase()) {
        return msg.send(`I am ${output}`);
      } else {
        return msg.send([phrase.name, tidbit.verb, output].join(' '));
      }
    }
    alias (msg) {
      let srcName = msg.match[1].trim();
      let targetName = msg.match[2].trim();
      let src = robot.phrase.get(srcName);
      if (src) {
        msg.reply(`Sorry, there is already a phrase for '${srcName}'.`);
        return;
      }
      let target = robot.phrase.get(targetName);
      if (!target) {
        msg.reply(`Sorry, there is no phrase for the target '${targetName}'.`);
        return;
      }
      if (!target.canAlias(msg.message.user)) {
        robot.logger.debug(`${phrase} that phrase is protected`);
        msg.reply('Sorry, that phrase is protected');
        return;
      }
      msg.finish();
      robot.logger.info(`${msg.message.user.name} aliased '${srcName}' to '${targetName}'`);
      var phrase = new Factoid(srcName);
      phrase.creator = msg.message.user.name;
      phrase.room = msg.message.user.room;
      phrase.alias = targetName;
      phrase.save();
      msg.reply('Okay.');
    }
    setAddressed (msg) {
      msg.message.addressed = true;
      return this.set(msg);
    }
    set (msg) {
      msg.finish(); // we are adding a message, so ignore any other handler type
      let fact = msg.match[1].trim();
      let verb = msg.match[2].trim();
      let tidbit = msg.match[3].trim();
      let forced = !!msg.match[4];
      if (!msg.message.addressed && /^[^a-zA-Z]*<.?\S+>/.test(fact)) {
        robot.logger.debug('Not learning from what seems to be an IRC quote: $fact');
        return;
      }
      if (!msg.message.addressed && !forced && /=~/.test(fact)) {
        robot.logger.debug('Not learning what looks like a botched =~ query');
        msg.reply('Fix your =~ command.');
        return;
      }

      if (fact === 'you' && verb === 'are') {
        fact = robot.name;
        verb = 'is';
      } else if (fact === 'I' && verb === 'am') {
        fact = msg.message.user.name;
        verb = 'is';
      }

      this.stats.learn++;
      let matches = tidbit.match(/^<(action|reply)>\s*(.*)/);
      if (matches) {
        verb = `<${matches[1]}>`;
        tidbit = matches[2];
      } else if (verb === 'is also') {
        verb = 'is';
      } else if (forced) {
        if (verb !== '<action>' && verb !== '<reply>') {
          verb = verb.replace(/^<|>$/, '');
        }
        if (!/\sis\salso$/.test(fact)) {
          fact.replace(/\sis$/, '');
        }
      }

      if (verb.toLowerCase === '<alias>') {
        msg.reply("please use the 'alias' command.");
        return;
      }

      fact = fact.trim();
      robot.logger.debug(`Learning ${fact} ${verb} ${tidbit}`);

      if (fact.toLowerCase() === msg.message.user.name.toLowerCase() || fact.toLowerCase() === msg.message.user.name.toLowerCase() + ' quotes') {
        robot.logger.debug(`Not allowing ${msg.message.user.name} to edit his own phrase`);
        msg.reply("Please don't edit your own phrases.");
        return;
      }

      let phrase = this.get(fact);
      if (!phrase) {
        phrase = new Factoid(fact);
        phrase.creator = msg.message.user.name;
        phrase.room = msg.message.user.room;
      } else if (!phrase.canEdit(msg.message.user)) {
        robot.logger.debug(`${phrase} that phrase is protected`);
        msg.reply('Sorry, that phrase is protected');
        return;
      }

      for (let t of phrase.tidbits) {
        if (tidbit.toLowerCase() === t.tidbit.toLowerCase()) {
          msg.reply('I already had it that way');
          return;
        }
      }

      phrase.tidbits.push({
        tidbit: tidbit,
        verb: verb,
        creator: msg.message.user.name,
        room: msg.message.user.room
      });
      phrase.save();
      robot.logger.debug(`${msg.message.user.name} taught in ${msg.message.user.room} ${phrase.tidbits.length} '${fact}', '${verb}' '${tidbit}'`);
      return msg.reply('Okay.');
    }
    handlerLiteral (msg) {
      // page - http://wiki.xkcd.com/irc/bucket#Listing_phrases
      let phraseName = msg.match[2].trim();
      let phrase = robot.phrase.get(phraseName);
      if (!phrase) {
        msg.reply('No such phrase');
        return;
      }
      msg.finish();
      if (phrase.tidbits.length > 10) {
        let baseurl = (process.env.HUBOT_URL || (`http://${os.hostname()}:${robot.server.address().port}`)).replace(/\/+$/, '');
        msg.reply(baseurl + '/' + robot.name + '/phrase/' + encodeURIComponent(phraseName));
        return;
      }
      let response = [];
      phrase.tidbits.forEach(tidbit => response.push(`${tidbit.verb} ${tidbit.tidbit}`));
      return msg.reply(`${phrase.name} (${phrase.tidbits.length}): ${response.join('|')}`);
    }
    handlerGetAddressed (msg) {
      msg.message.addressed = true;
      return this.handlerGet(msg);
    }
    handlerGet (msg) {
      let phraseName = msg.match[1].trim().replace(/[.,/#!$%^&*;:{}=-_`~()]/g, '');
      // FIXME this should be per room
      let history = [];
      let phrase = robot.phrase.get(phraseName, history);
      if (!phrase) { return; }
      robot.phrase.output(msg, phrase, history);
      msg.finish();
      if (history.length > 0) {
        robot.phrase.last_phrase = history;
      }
    }
  }

  robot.phrase = new FactoidHandler();
  robot.phrase.last_phrase = null;

  robot.respond(/(?:do something|something random)$/, msg => {
    let history = [];
    let phrase = robot.phrase.random(history);
    robot.phrase.output(msg, phrase, history);
    if (history.length > 0) { robot.phrase.last_phrase = history; }
  });

  robot.hear(/^(?:do something|something random)$/, msg => {
    let history = [];
    let phrase = robot.phrase.random(history);
    robot.phrase.output(msg, phrase, history);
    if (history.length > 0) { robot.phrase.last_phrase = history; }
  });

  robot.respond(/(un)?protect\s*(.*)$/, msg => {
    let protect = !msg.match[1];
    let phraseName = msg.match[2].trim();
    let phrase = robot.phrase.get(phraseName);
    if (!phrase) {
      msg.reply('No such phrase.');
      return;
    }
    msg.finish();
    if (phrase.readonly === protect) {
      msg.reply('I already had it that way.');
      return;
    }
    phrase.readonly = protect;
    phrase.save();
    return msg.reply('Okay.');
  });

  robot.respond(/alias (.*?) => (.*?)$/, robot.phrase.alias);

  robot.router.get(`/${robot.name}/phrase/:phrase`, function (req, res) {
    let phraseName = req.params.phrase;
    let phrase = robot.phrase.get(phraseName);
    if (!phrase) { return res.status(404).send('Not Found'); }
    res.setHeader('content-type', 'text/plain');
    let content = [];
    content.push(`Factoid: [${phraseName}]`);
    content.push(`Protected: ${phrase.readonly ? 'true' : 'false'}`);
    content.push('');
    content.push('Tidbits:');
    for (let tidbit of phrase.tidbits) {
      content.push(tidbit.verb + '|' + tidbit.tidbit);
    }
    res.send(content.join('\n'));
    return res.end;
  });
    // res.send require('util').inspect(req.params)

  robot.respond(/literal(?:\[([*\d]+)])?\s+(.*)$/, robot.phrase.handlerLiteral);
  robot.hear(/^literal(?:\[([*\d]+)])?\s+(.*)$/, robot.phrase.handlerLiteral);

  robot.respond(/forget #(\d+)$/, msg => {
    return msg.reply('Sorry, syntax is now "forget <phrase>#<index from 0>" or "forget that"');
  });

  robot.respond(/forget that$/, msg => {
    return msg.reply('FIXME - not implemented yet');
  });

  robot.respond(/forget (.+)#(\d+)$/, msg => {
    let phraseName = msg.match[1].trim();
    let tid = parseInt(msg.match[2], 10) - 1;
    let phrase = robot.phrase.get(phraseName);
    if (!phrase) {
      msg.reply('No such phrase');
      return;
    }
    if (!phrase.canEdit(msg.message.user)) {
      msg.reply(`Sorry, you don't have permissions to edit '${phrase.name}'.`);
      return;
    }
    if (tid < 0) {
      msg.reply('Sorry, you must provide a number greater than 0 (as this is 1 based)');
      return;
    }
    if (!phrase.tidbits[tid]) {
      msg.reply(`Can't find tidbit #${tid + 1}`);
      return;
    }
    let tidbit = phrase.tidbits.splice(tid, 1);
    msg.reply(`Deleted tidbit: ${tidbit[0].verb}|${tidbit[0].tidbit}`);
    return phrase.save();
  });

  robot.respond(/what was that\??$/, function (msg) {
    // FIXME this should be per room
    if (!robot.phrase.last_phrase) { return; }
    msg.finish();
    // halkeye: That was 'rofl' (#315): <reply> I am also amused
    // halkeye: That was 'that's what she said' => 'thats what she said' (#65): <reply> No, that's what HE said.
    // halkeye: That was 'give me a weapon' (#863): <action> gives $weapon to $who;  vars used: { 'weapon' => [ 'a Biggoron Sword' ]};.
    let lf = robot.phrase.last_phrase.slice(-1)[0];
    let { name } = lf.phrase;
    let { tidbit } = lf;
    let idx = lf.phrase.tidbits.map(tid => tid.tidbit).indexOf(tidbit.tidbit);

    let response = [];
    response.push('That was');
    if (robot.phrase.last_phrase.length > 2) {
      robot.phrase.last_phrase.slice(0, -2).forEach(fact => response.push(`'${fact.name}' =>`));
    }
    response.push(`'${name}'`);
    response.push(`(#${idx}):`);
    response.push(tidbit.verb);
    response.push(tidbit.tidbit);
    if (lf.vars) {
      response.push(';');
      response.push('vars used:');
      response.push(util.inspect(lf.vars, { depth: null }));
    }
    return msg.reply(response.join(' '));
  });

  robot.respond(/(.*?)\s+(<\w+(?:'t)?>)\s*(.*)()/i, robot.phrase.setAddressed);
  robot.respond(/(.*?)(<'s>)\s+(.*)()/i, robot.phrase.setAddressed);
  robot.respond(/(.*?)\s+(is(?: also)?|are)\s+(.*)/i, robot.phrase.setAddressed);
  robot.hear(/^(.*?)\s+(<\w+(?:'t)?>)\s*(.*)()/i, robot.phrase.set);
  robot.hear(/^(.*?)(<'s>)\s+(.*)()/i, robot.phrase.set);
  robot.hear(/^(.*?)\s+(is(?: also)?|are)\s+(.*)/i, robot.phrase.set);

  // # FIXME, make these loaded once brain is loaded so it doesn't need to do wildcard match
  robot.respond(/(.{3,})$/, robot.phrase.handlerGetAddressed);
  robot.hear(/(.{3,})$/, robot.phrase.handlerGet);
};

